src/client/__tests__/calculate-updates.spec.js

```diff
@@ -0,0 +1,231 @@
+		expect( updates ).toEqual( [
+			{ endpointPath: [ 'things' ], params: { page: 1, perPage: 3 } },
+		] );
+	} );
+
+	it( 'should return an empty set if nothing needs an update', () => {
+		const requirementsByEndpoint = {
+			things: {
+				endpoints: {
+					1: { freshness: 60 * SECOND },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, freshness: 120 * SECOND },
+				],
+			},
+		};
+		const endpointsState = {
+			things: {
+				endpoints: {
+					1: { lastReceived: now - ( 59 * SECOND ) },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, lastReceived: now - ( 50 * SECOND ) },
+				],
+			},
+		};
+
+		const { updates } = calculateUpdates( requirementsByEndpoint, endpointsState, now );
+		expect( updates ).toEqual( [] );
+	} );
+
+	it( 'should add updates for requirements that are not yet present in state', () => {
+		const requirementsByEndpoint = {
+			things: {
+				endpoints: {
+					1: { freshness: 60 * SECOND },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, freshness: 120 * SECOND },
+				],
+			},
+		};
+
+		const { updates } = calculateUpdates( requirementsByEndpoint, {}, now );
+		expect( updates ).toEqual( [
+			{ endpointPath: [ 'things', '1' ] },
+			{ endpointPath: [ 'things' ], params: { page: 1, perPage: 3 } },
+		] );
+	} );
+
+	it( 'should give a default nextUpdate value in the absence of any needed updates', () => {
+		expect( calculateUpdates( {}, {} ).nextUpdate ).toEqual( DEFAULT_NEXT_UPDATE );
+	} );
+
+	it( 'should return default value if nothing is required earlier.', () => {
+		const requirementsByEndpoint = {
+			things: {
+				endpoints: {
+					1: { freshness: DEFAULT_NEXT_UPDATE + SECOND },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, freshness: DEFAULT_NEXT_UPDATE + SECOND },
+				],
+			},
+		};
+		const endpointsState = {
+			things: {
+				endpoints: {
+					1: { lastReceived: now - ( 1 * SECOND ) },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, lastReceived: now - ( 1 * SECOND ) },
+				],
+			},
+		};
+		const { nextUpdate } = calculateUpdates( requirementsByEndpoint, endpointsState );
+		expect( nextUpdate / 1000 ).toBeCloseTo( DEFAULT_NEXT_UPDATE / 1000, 1 );
+	} );
+
+	it( 'should return time difference for earliest required endpoint.', () => {
+		const requirementsByEndpoint = {
+			things: {
+				endpoints: {
+					1: { freshness: 60 * SECOND },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, freshness: 90 * SECOND },
+				],
+			},
+		};
+		const endpointsState = {
+			things: {
+				endpoints: {
+					1: { lastReceived: now - ( 40 * SECOND ) },
+				},
+				queries: [
+					{ params: { page: 1, perPage: 3 }, lastReceived: now - ( 40 * SECOND ) },
+				],
+			},
+		};
+		const { nextUpdate } = calculateUpdates( requirementsByEndpoint, endpointsState );
+		expect( nextUpdate / 1000 ).toBeCloseTo( ( 20 * SECOND ) / 1000, 1 );
```

Comment:
TIL `toBeCloseTo`

Reply:
Yeah, these tests were giving me fits as the date math was barely imprecise. Probably using floating point math internally even though I'm dealing in integers. That's why I did the `/ 1000`.

I figure if they're accurate to < 1 second, that's good enough for our purposes here.